packages/flutter/lib/src/widgets/viewport.dart

```diff
@@ -219,26 +222,64 @@ class _ViewportElement extends MultiChildRenderObjectElement {
 
   @override
   void mount(Element? parent, Object? newSlot) {
+    assert(!_doingMountOrUpdate);
+    _doingMountOrUpdate = true;
     super.mount(parent, newSlot);
     _updateCenter();
+    assert(_doingMountOrUpdate);
+    _doingMountOrUpdate = false;
   }
 
   @override
   void update(MultiChildRenderObjectWidget newWidget) {
+    assert(!_doingMountOrUpdate);
+    _doingMountOrUpdate = true;
     super.update(newWidget);
     _updateCenter();
+    assert(_doingMountOrUpdate);
+    _doingMountOrUpdate = false;
   }
 
   void _updateCenter() {
     // TODO(ianh): cache the keys to make this faster
     if (widget.center != null) {
-      renderObject.center = children.singleWhere(
-        (Element element) => element.widget.key == widget.center,
-      ).renderObject as RenderSliver?;
+      for (int i = 0; i < children.length; i++) {
+        if (children.elementAt(i).widget.key == widget.center) {
+          renderObject.center = children.elementAt(i).renderObject as RenderSliver?;
+          _centerSlotIndex = i;
+          break;
+        }
+      }
     } else if (children.isNotEmpty) {
       renderObject.center = children.first.renderObject as RenderSliver?;
+      _centerSlotIndex = 0;
     } else {
       renderObject.center = null;
+      _centerSlotIndex = null;
+    }
+  }
+
+  @override
+  void insertRenderObjectChild(RenderObject child, IndexedSlot<Element?> slot) {
+    super.insertRenderObjectChild(child, slot);
+    // Once [mount]/[update] are done, the `renderObject.center` will be updated
+    // in [_updateCenter].
+    if (!_doingMountOrUpdate) {
+      if (slot.index == _centerSlotIndex) {
+        renderObject.center = child as RenderSliver?;
+      }
+    }
+  }
+
+  // `moveRenderObjectChild` only happens during [update] stage.
```

Comment:
Seems an invalid comment that's indicating the wrong method and didn't start with `///`?

Reply:
Just comment that why not override `moveRenderObjectChild`

Reply:
Could it say it? It wasn't immediately clear for me either
For example

```
// We don't override `moveRenderObjectChild` here, because it only happens
// during [update] stage.
```

Reply:
Or: Do override moveRenderObjectChild, which just calls super and has an assert that _doingMountOrUpdate is true to show & proof that the comment is true.

Reply:
> Or: Do override moveRenderObjectChild, which just calls super and has an assert that _doingMountOrUpdate is true to show & proof that the comment is true.

Done and add a test to test it.